### PR TITLE
[Source] Add a SourceFileManager to lazily load file contents

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

File: Sources/ClangImporter/ClangImporter.swift [hunk 1]
```diff
@@ -345,7 +314,7 @@ public class ClangImporter: Pass {
       let name = Identifier(name: clang_getCursorSpelling(child).asSwift())
       if context.global(named: name) != nil { return CXChildVisit_Continue }
 
-      let range = SourceRange(clangRange: clang_getCursorExtent(child))
+      let range = self.self.makeRange(clang_getCursorExtent(child))
```
Comment (harlanhaskins) [inline Sources/ClangImporter/ClangImporter.swift]: self.self.?

File: Sources/Source/SourceFileManager.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
+///
+/// SourceFile.swift
```
Comment (harlanhaskins) [inline Sources/Source/SourceFileManager.swift]: SourceFileManager.swift

File: Sources/Source/SourceFileManager.swift [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+/// Full license text available at https://github.com/trill-lang/trill
+///
+
+public final class SourceFileManager {
+  private var contentsCache = [SourceFile: String]()
+  private var linesCache = [SourceFile: [String]]()
+
+  public init() {}
+
+  public func contents(file: SourceFile) throws -> String {
```
Comment (harlanhaskins) [inline Sources/Source/SourceFileManager.swift]: contents(of:) I think?

File: Sources/Source/SourceLocation.swift [hunk 4]
```diff
@@ -55,3 +61,17 @@ extension SourceRange: Equatable {
...
+extension SourceRange {
+  private var contents: String { return start.file.contents }
+
+  public var length: Int {
+    return end.charOffset - start.charOffset
+  }
+
+  public var source: String {
+    let startIndex = contents.index(contents.startIndex, offsetBy: start.charOffset)
+    let endIndex = contents.index(contents.startIndex, offsetBy: end.charOffset)
```
Comment (harlanhaskins) [inline Sources/Source/SourceLocation.swift]: pass startIndex as the first argument and use the difference of the offsets as the offset from startIndex to the end. That should save a bunch of work.

**Review by harlanhaskins — COMMENTED** [review]
LGTM modulo nitpicks

File: Sources/ClangImporter/ClangImporter.swift (hunk 1 above)
Comment (segiddins) [reply Sources/ClangImporter/ClangImporter.swift]: oops, accidentally applied a fix-it after I'd fixed it

**Pull request merged**